File: Source/TWXProxy-MB/Encryptor.pas
{
Copyright (C) 2005  Remco Mulder

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

For source notes please refer to Notes.txt
For license terms please refer to GPL.txt.

These files should be stored in the root of the compression you 
received this source in.
}
unit
  Encryptor;

interface

uses
  Classes, Sysutils;

type
  EEncryptError = class(Exception);

  TEncryptor = class(TComponent)
  protected
    { Protected declarations }
    FKey                : String;
    FShift              : Integer;
    FShiftKey,
    FChunkSize,
    FScrambleSeed       : Byte;
  public
    { Public declarations }
    procedure Encrypt(var Target : String);
    procedure Decrypt(var Target : String);
    procedure ConvertKey(var Key : String);
  published
    { Published declarations }
    property Key : String Read FKey Write FKey;
    property Shift : Integer Read FShift Write FShift default 5;
    property ShiftKey : Byte Read FShiftKey Write FShiftKey default 50;
    property ChunkSize : Byte Read FChunkSize Write FChunkSize default 18;
    property ChunkKey : Byte Read FScrambleSeed Write FScrambleSeed;
  end;

procedure Register;

implementation

procedure Register;
begin
  RegisterComponents('Samples', [TEncryptor]);
end;

procedure TEncryptor.ConvertKey(var Key : String);
var
  S       : String;
  I       : Integer;
  KeyChar : String;
begin
  KeyChar := '';

  for I := 1 to Length(Key) do
  begin
    if (Key[I] = ',') then
    begin
      try
        S := S + Char(StrToInt(KeyChar));
        KeyChar := '';
      except
        raise EEncryptError.Create('Bad encryption key format');
      end;
    end
    else
      KeyChar := KeyChar + Key[I];
  end;

  try
    S := S + Char(StrToInt(KeyChar));
  except
    raise EEncryptError.Create('Bad encryption key format');
  end;

  Key := S;
end;

procedure TEncryptor.Encrypt(var Target : String);
var
  Chunks,
  I,
  X,
  ChunkStart : Integer;
  J,
  B,
  C,
  ChunkKey,
  Last,
  CheckSum   : Byte;
  S,
  Key,
  Chunk      : String;
  P          : ^Char;
begin
  // Encode string
  S := '';
  Last := $18;
  CheckSum := $F0;
  Key := FKey;
  ConvertKey(Key);

  for I := 1 to Length(Target) do
  begin
    B := Ord(Target[I]);
    CheckSum := CheckSum xor B;
    C := B;

    for X := 1 to Length(Key) do
      B := B xor Ord(Key[X]);

    if (I mod FShift = 0) then
      // apply shift
      B := B xor FShiftKey;

    B := B xor Last;

    Last := C;
    S := S + Chr(B);
  end;

  // add checksum
  S := S + Char(CheckSum);

  // Scramble string
  Target := '';
  Chunks := Length(S) DIV FChunkSize;

  if (Chunks * FChunkSize < Length(S)) then
    Chunks := Chunks + 1;

  for I := 1 to Chunks do
  begin
    ChunkStart := (I - 1) * FChunkSize + 1;

    if (ChunkStart + FChunkSize - 1 > Length(S)) then
      X := Length(S) - ChunkStart + 1
    else
      X := FChunkSize;

    Chunk := Copy(S, ChunkStart, X);
    ChunkKey := Byte(Chunk[1]);

    // record index of chunk and encrypt it -
    // do this by accessing 32-bit chunk index in memory and
    // encrypting it at low level
    P := @I;
    for J := 1 to SizeOf(Integer) do
    begin
      Chunk := Char(Byte(P^) xor FScrambleSeed xor ChunkKey xor J) + Chunk;
      P := Pointer(Integer(P) + 1);
    end;

    if (Random < 0.5) or (X < FChunkSize) then
      Target := Target + Chunk
    else
      Target := Chunk + Target;
  end;
end;

procedure TEncryptor.Decrypt(var Target : String);
var
  I,
  X,
  ChunkIndex : Integer;
  J,
  B,
  ChunkKey,
  CheckSum,
  Last       : Byte;
  S,
  Chunk,
  Key,
  ChunkIdx   : String;
  P          : Pointer;
begin
  // Unscramble string
  I := 1;
  X := 1;
  Key := FKey;
  ConvertKey(Key);

  while (I < Length(Target)) do
  begin
    // search for a chunk with the index we're looking for (X)

    // get this chunk
    Chunk := Copy(Target, I, FChunkSize + SizeOf(Integer));

    if (Length(Chunk) < SizeOf(Integer) + 1) then
      // invalid chunk - unscramble failed
      Raise EEncryptError.Create('Decryption failure');

    // decode the chunk index
    ChunkKey := Byte(Chunk[5]);
    P := PChar(Chunk);
    ChunkIdx := '';
    for J := 1 to SizeOf(Integer) do
    begin
      Byte(P^) := Byte(P^) xor (5 - J) xor ChunkKey xor FScrambleSeed;
      ChunkIdx := Char(P^) + ChunkIdx;
      P := Pointer(Integer(P) + 1);
    end;

    ChunkIndex := Integer(Pointer(ChunkIdx)^);

    if (ChunkIndex = X) then
    begin
      S := S + Copy(Chunk, SizeOf(Integer) + 1, Length(Chunk));
      I := 1;
      Inc(X);
    end
    else
      I := I + Length(Chunk);
  end;

  if (S = '') then
    Raise EEncryptError.Create('Decryption failure');

  // Decode string
  Last := $18;
  Target := '';
  CheckSum := $F0;

  for I := 1 to Length(S) - 1 do
  begin
    B := Ord(S[I]);

    B := B xor Last;

    if (I mod FShift = 0) then
      // apply shift
      B := B xor FShiftKey;

    for X := Length(Key) downto 1 do
      B := B xor Ord(Key[X]);

    Target := Target + Char(B);
    CheckSum := CheckSum xor B;
    Last := B;
  end;

  if (CheckSum <> Byte(S[Length(S)])) then
    Raise EEncryptError.Create('Decryption failure');
end;

end.

File: Source/TWXProxy-C#/Encryptor.cs
using System;
using System.ComponentModel;
 // 
 // Copyright (C) 2005  Remco Mulder
 // 
 // This program is free software; you can redistribute it and/or modify
 // it under the terms of the GNU General Public License as published by
 // the Free Software Foundation; either version 2 of the License, or
 // (at your option) any later version.
 // 
 // This program is distributed in the hope that it will be useful,
 // but WITHOUT ANY WARRANTY; without even the implied warranty of
 // MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 // GNU General Public License for more details.
 // 
 // You should have received a copy of the GNU General Public License
 // along with this program; if not, write to the Free Software
 // Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 // 
 // For source notes please refer to Notes.txt
 // For license terms please refer to GPL.txt.
 // 
 // These files should be stored in the root of the compression you
 // received this source in.
namespace Encryptor
{
    public class EEncryptError: Exception
    {
        //@ Constructor auto-generated 
        public EEncryptError(String message)
            :base(message)
        {
        }
        //@ Constructor auto-generated 
        public EEncryptError(String message, Exception innerException)
            :base(message, innerException)
        {
        }
    } // end EEncryptError

    public class TEncryptor: Component
    {
        // Published declarations
        public string Key
        {
          get {
            return FKey;
          }
          set {
            FKey = value;
          }
        }
        public int Shift
        {
          get {
            return FShift;
          }
          set {
            FShift = value;
          }
        }
        public byte ShiftKey
        {
          get {
            return FShiftKey;
          }
          set {
            FShiftKey = value;
          }
        }
        public byte ChunkSize
        {
          get {
            return FChunkSize;
          }
          set {
            FChunkSize = value;
          }
        }
        public byte ChunkKey
        {
          get {
            return FScrambleSeed;
          }
          set {
            FScrambleSeed = value;
          }
        }
        // Protected declarations
        protected string FKey = String.Empty;
        protected int FShift = 5;
        protected byte FShiftKey = 50;
        protected byte FChunkSize = 18;
        protected byte FScrambleSeed = 0;
        public void ConvertKey(ref string Key)
        {
            string S;
            int I;
            string KeyChar;
            KeyChar = "";
            for (I = 1; I <= Key.Length; I ++ )
            {
                if ((Key[I] == ','))
                {
                    try {
                        S = S + ((char)Convert.ToInt32(KeyChar));
                        KeyChar = "";
                    }
                    catch {
                        throw new EEncryptError("Bad encryption key format");
                    }
                }
                else
                {
                    KeyChar = KeyChar + Key[I];
                }
            }
            try {
                S = S + ((char)Convert.ToInt32(KeyChar));
            }
            catch {
                throw new EEncryptError("Bad encryption key format");
            }
            Key = S;
        }

        // Public declarations
        public void Encrypt(ref string Target)
        {
            int Chunks;
            int I;
            int X;
            int ChunkStart;
            byte J;
            byte B;
            byte C;
            byte ChunkKey;
            byte Last;
            byte CheckSum;
            string S;
            string Key;
            string Chunk;
            char P;
            // Encode string
            S = "";
            Last = 0x18;
            CheckSum = 0xF0;
            Key = FKey;
            ConvertKey(ref Key);
            for (I = 1; I <= Target.Length; I ++ )
            {
                B = (int)(Target[I]);
                CheckSum = CheckSum ^ B;
                C = B;
                for (X = 1; X <= Key.Length; X ++ )
                {
                    B = B ^ (int)(Key[X]);
                }
                if ((I % FShift == 0))
                {
                    // apply shift
                    B = B ^ FShiftKey;
                }
                B = B ^ Last;
                Last = C;
                S = S + (char)(B);
            }
            // add checksum
            S = S + (char)CheckSum;
            // Scramble string
            Target = "";
            Chunks = S.Length / FChunkSize;
            if ((Chunks * FChunkSize < S.Length))
            {
                Chunks = Chunks + 1;
            }
            for (I = 1; I <= Chunks; I ++ )
            {
                ChunkStart = (I - 1) * FChunkSize + 1;
                if ((ChunkStart + FChunkSize - 1 > S.Length))
                {
                    X = S.Length - ChunkStart + 1;
                }
                else
                {
                    X = FChunkSize;
                }
                Chunk = S.Substring(ChunkStart - 1 ,X);
                ChunkKey = ((byte)Chunk[1]);
                // record index of chunk and encrypt it -
                // do this by accessing 32-bit chunk index in memory and
                // encrypting it at low level
                P = I;
                for (J = 1; J <= sizeof(int); J ++ )
                {
                    Chunk = ((char)(byte)P ^ FScrambleSeed ^ ChunkKey ^ J) + Chunk;
                    P = (P + 1 as object);
                }
                if (((new System.Random()).NextDouble() < 0.5) || (X < FChunkSize))
                {
                    Target = Target + Chunk;
                }
                else
                {
                    Target = Chunk + Target;
                }
            }
        }

        public void Decrypt(ref string Target)
        {
            int I;
            int X;
            int ChunkIndex;
            byte J;
            byte B;
            byte ChunkKey;
            byte CheckSum;
            byte Last;
            string S;
            string Chunk;
            string Key;
            string ChunkIdx;
            object P;
            // Unscramble string
            I = 1;
            X = 1;
            Key = FKey;
            ConvertKey(ref Key);
            while ((I < Target.Length))
            {
                // search for a chunk with the index we're looking for (X)
                // get this chunk
                Chunk = Target.Substring(I - 1 ,FChunkSize + sizeof(int));
                if ((Chunk.Length < sizeof(int) + 1))
                {
                    // invalid chunk - unscramble failed
                    throw new EEncryptError("Decryption failure");
                }
                // decode the chunk index
                ChunkKey = ((byte)Chunk[5]);
                P = (Chunk as string);
                ChunkIdx = "";
                for (J = 1; J <= sizeof(int); J ++ )
                {
                    (byte)P = (byte)P ^ (5 - J) ^ ChunkKey ^ FScrambleSeed;
                    ChunkIdx = (char)P + ChunkIdx;
                    P = ((int)P + 1 as object);
                }
                ChunkIndex = (ChunkIdx as object);
                if ((ChunkIndex == X))
                {
                    S = S + Chunk.Substring(sizeof(int) + 1 - 1 ,Chunk.Length);
                    I = 1;
                    X ++;
                }
                else
                {
                    I = I + Chunk.Length;
                }
            }
            if ((S == ""))
            {
                throw new EEncryptError("Decryption failure");
            }
            // Decode string
            Last = 0x18;
            Target = "";
            CheckSum = 0xF0;
            for (I = 1; I < S.Length; I ++ )
            {
                B = (int)(S[I]);
                B = B ^ Last;
                if ((I % FShift == 0))
                {
                    // apply shift
                    B = B ^ FShiftKey;
                }
                for (X = Key.Length; X >= 1; X-- )
                {
                    B = B ^ (int)(Key[X]);
                }
                Target = Target + (char)B;
                CheckSum = CheckSum ^ B;
                Last = B;
            }
            if ((CheckSum != ((byte)S[S.Length])))
            {
                throw new EEncryptError("Decryption failure");
            }
        }

    } // end TEncryptor

}

namespace Encryptor.Units
{
    public class Encryptor
    {
        public static void Register()
        {
            //@ Unsupported function or procedure: 'RegisterComponents'
            RegisterComponents("Samples", new TEncryptor[] {TEncryptor});
        }

    } // end Encryptor

}

